### PR TITLE
Fireobj Kinda OK

### PR DESCRIPTION
**Pull request by louist103**
Before opening this PR, ensure the following:
- `./format.sh` was run to apply standard formatting.
- `make` successfully builds a matching ROM.
- No new compiler warnings were introduced during the build process.
    - Can be verified locally by running `tools/warnings_count/check_new_warnings.sh`
- New variables & functions should follow standard naming conventions.
- Comments and variables have correct spelling.
---

All function instructions match but data is accessed incorrectly.

### CODE REVIEW

File: src/code/z_fireobj.c [hunk 1]
```diff
@@ -1,27 +1,244 @@
 #include "global.h"
+#include "overlays/actors/ovl_En_Arrow/z_en_arrow.h"
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_fireobj/FireObj_InitWithParams.s")
+ColliderCylinderInit fireObjCollisionInit = {
```
Comment (Thar0) [inline src/code/z_fireobj.c:4]:
```suggestion
ColliderCylinderInit sFireObjCylinderInit = {
```

File: src/code/z_fireobj.c [hunk 2]
```diff
@@ -1,27 +1,244 @@
...
+    fire->yScale = 0.0f;
+    fire->dynamicSize = 0.0f;
+    fire->unk26 = Rand_ZeroOne() * 20.0f;
+    fire->ignitionDelay = -1;
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_fireobj/FireObj_UpdateStateTransitions.s")
+void FireObj_SetState(FireObj* fire, f32 dynamicSizeStep, u8 newState) {
+    fire->state = newState & 0xFF;
+    if (fire->state == 3) {
```
Comment (Thar0) [inline src/code/z_fireobj.c:49]: States should have an enum. For naming the values it seems like there may be enough in this file to surmise what the different states do, but if not even an enum like `FIREOBJ_STATE_0` etc. is better than nothing.

File: src/code/z_fireobj.c [hunk 3]
```diff
@@ -1,27 +1,244 @@
...
+        gSPSegment(POLY_XLU_DISP++, 0x08,
+                   Gfx_TwoTexScroll(globalCtx->state.gfxCtx, 0, 0, 0, 0x20, 0x40, 1, 0, (fire->unk26 * -0x14) & 0x1FF,
+                                    0x20, 0x80));
+
+        gDPSetPrimColor(POLY_XLU_DISP++, 0, temp_v1->unk4, temp_v1->unk0, temp_v1->unk1, temp_v1->unk2, temp_v1->unk3);
+
+        gDPSetEnvColor(POLY_XLU_DISP++, temp_v1->unk5, temp_v1->unk6, temp_v1->unk7, 0);
+
+        vec.x = 0;
+        vec.y = func_800DFCDC(globalCtx->cameraPtrs[globalCtx->activeCamera]) + 0x8000;
```
Comment (Thar0) [inline src/code/z_fireobj.c:150]:
```suggestion
        vec.y = func_800DFCDC(GET_ACTIVE_CAM(globalCtx)) + 0x8000;
```

File: src/code/z_fireobj.c [hunk 4]
```diff
@@ -1,27 +1,244 @@
...
+        vec.z = 0;
+        Matrix_SetStateRotationAndTranslation(fire->position.x, fire->position.y, fire->position.z, &vec);
+        Matrix_Scale(fire->xScale, fire->yScale, 1.0f, 1);
```
Comment (Thar0) [inline src/code/z_fireobj.c:153]:
```suggestion
        Matrix_Scale(fire->xScale, fire->yScale, 1.0f, MTXMODE_APPLY);
```

File: src/code/z_fireobj.c [hunk 5]
```diff
@@ -1,27 +1,244 @@
...
+    if (fire->state == 3) {
+        Lights_PointSetColorAndRadius((LightInfo*)&light->lightInfo, 0U, 0U, 0U, (s16)-1);
+    } else {
+        radius = ((fire->yScale * 140.0f * fire->sizeInv) + 60.0f);
+
+        Lights_PointGlowSetInfo((LightInfo*)&light->lightInfo, fire->position.x,
+                                (fire->position.y + (fire->yScale * 6500.0f)), fire->position.z,
+                                ((s32)(Rand_ZeroOne() * lightParams->unk5) + lightParams->unk2),
+                                ((s32)(Rand_ZeroOne() * lightParams->unk6) + lightParams->unk3),
+                                ((s32)(Rand_ZeroOne() * lightParams->unk7) + lightParams->unk4), radius);
```
Comment (Thar0) [inline src/code/z_fireobj.c:196]: It should be pretty clear what most or all of the `FireObjLightParams` fields are for now.

File: src/code/z_fireobj.c [hunk 6]
```diff
@@ -1,27 +1,244 @@
...
+        BUMP_ON,
+        OCELEM_NONE,
+    },
+    { 18, 67, 0, { 0, 0, 0 } },
+};
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_fireobj/FireObj_SetState.s")
+FireObjLightParams sFireObjLightParams[2] = {
+    { 0xFF, 0xFF, 0x00, 0xFF, 0x80, 0xFF, 0x00, 0x00 },
+    { 0x00, 0xC8, 0x80, 0x80, 0x00, 0x7F, 0x7F, 0x00 },
```
Comment (Thar0) [inline src/code/z_fireobj.c:26]: Some (maybe all?) of these fields are colors and so should be decimal.

File: src/code/z_fireobj.c [hunk 7]
```diff
@@ -1,27 +1,244 @@
...
+    s32 pad;
+    FireObjLightParams* temp_v1 = &sFireObjLightParams[fire->unk27];
+
+    if (fire->state != 3) {
+        Vec3s vec;
+        OPEN_DISPS(globalCtx->state.gfxCtx);
+        func_8012C2DC(globalCtx->state.gfxCtx);
+        gSPSegment(POLY_XLU_DISP++, 0x08,
+                   Gfx_TwoTexScroll(globalCtx->state.gfxCtx, 0, 0, 0, 0x20, 0x40, 1, 0, (fire->unk26 * -0x14) & 0x1FF,
+                                    0x20, 0x80));
```
Comment (Thar0) [inline src/code/z_fireobj.c:143]:
```suggestion
                   Gfx_TwoTexScroll(globalCtx->state.gfxCtx, 0, 0, 0, 32, 64, 1, 0, (fire->unk26 * -20) & 0x1FF,
                                    32, 128));
```

File: src/code/z_fireobj.c [hunk 8]
```diff
@@ -1,27 +1,244 @@
...
+        CLOSE_DISPS(globalCtx->state.gfxCtx);
+    }
+}
+
+#ifdef NON_EQUIVALENT
+// Accesses data incorrectly
+void FireObj_InitLight(GlobalContext* globalCtx, FireObjLight* light, u8* param_3, Vec3f* pos) {
+    FireObjLightParams* objectParams;
+
+    objectParams = &sFireObjLightParams[*param_3];
```
Comment (Thar0) [inline src/code/z_fireobj.c:166]: imo the code is cleaner this way but I don't think we have any particular preference, so this is optional.
```suggestion
    FireObjLightParams* objectParams = &sFireObjLightParams[*param_3];
```

**Review by Thar0 — CHANGES_REQUESTED** [review]
Some things

File: src/code/z_fireobj.c [hunk 9]
```diff
@@ -1,27 +1,244 @@
...
+    s32 sp40 = 0;
+    u8 nextState;
+    Vec3f dist;
+
+    FireObj_StepSize(fire);
+    fire->unk26 += 1;
+    if (fire->ignitionDelay > 0) {
+        fire->ignitionDelay--;
+    } else if (fire->ignitionDelay == 0) {
+        fire->ignitionDelay = -1;
```
[the remark this answers is not in the document]
Comment (louist103) [reply src/code/z_fireobj.c:100]: This isn't decrementing.  Its setting it to -1.

File: src/code/z_fireobj.c [hunk 10]
```diff
@@ -1,27 +1,251 @@
...
+        if ((fire->state == FIRE_STATE_3) || (fire->state == FIRE_STATE_1)) {
+            nextState = FIRE_STATE_0;
+        } else {
+            nextState = FIRE_STATE_1;
+        }
+        FireObj_SetState(fire, fire->dynamicSizeStep, nextState);
+    }
+    if ((fire->flags & 1) && (fire->state != FIRE_STATE_3) &&
+        (func_800CA1E8(globalCtx, &globalCtx->colCtx, fire->position.x, fire->position.z, &sp44, &waterBox) != 0) &&
+        ((fire->yScale * ((void)0, 6500.0f)) < (sp44 - fire->position.y))) { // Fake but IDK what else
```
Comment (jpburnett) [inline src/code/z_fireobj.c:117]:
```suggestion
        ((fire->yScale * (NULL, 6500.0f)) < (sp44 - fire->position.y))) { // Fake but IDK what else
```
could this be a Null pointer?

Comment (louist103) [reply src/code/z_fireobj.c:117]: None of the other instances of this trick use that.

Comment (AngheloAlf) [reply src/code/z_fireobj.c:117]: `NULL` usually expands to `(void*)0` (note the `*` which means it is a pointer), the `void` here just silences an "unused value" warning

File: tools/warnings_count/warnings_build_current.txt [hunk 11]
```diff
@@ -1,2 +1,6 @@
 cc: Warning: -mips3 should not be used for ucode 32-bit compiles
 cc: Warning: -mips3 should not be used for ucode 32-bit compiles
+[01m[Ksrc/code/z_fireobj.c:[m[K In function ‘[01m[KFireObj_UpdateStateTransitions[m[K’:
+[01m[Ksrc/code/z_fireobj.c:110:28:[m[K [01;35m[Kwarning: [m[Kleft-hand operand of comma expression has no effect [[01;35m[K-Wunused-value[m[K]
+  110 |         ((fire->yScale * (0[01;35m[K,[m[K 6500.0f)) < (sp44 - fire->position.y))) { // Fake but IDK what else
+      |                            [01;35m[K^[m[K
```
Comment (AngheloAlf) [inline tools/warnings_count/warnings_build_current.txt:6]: Does this warning still exist? If it doesn't, then remember to update this file (just ensuring this file isn't being updated should be enough)

Comment (louist103) [reply tools/warnings_count/warnings_build_current.txt:6]: I re ran the warnings script and I guess it didn't work.  is it OK to remove it manually?

Comment (AngheloAlf) [reply tools/warnings_count/warnings_build_current.txt:6]: Yeah. Worst case scenario Jenkins would be mad

**Review by AngheloAlf — COMMENTED** [review]
Did you run the formatter on z64.h? It doesn't look good...
Could you fix that?

**Review by Kenix3 — APPROVED** [review]

**Comment by louist103** [comment]
> Did you run the formatter on z64.h? It doesn't look good... Could you fix that?

I just ran format.sh. I can try and revert it.

File: tools/warnings_count/warnings_build_current.txt (hunk 11 above)
Comment (AngheloAlf) [reply tools/warnings_count/warnings_build_current.txt:6]: What happened with this?

**Review by AngheloAlf — APPROVED** [review]

**Pull request merged**